README.md
[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)
[![Donate](https://badgen.net/badge/donate/paypal)](https://paypal.me/HomebridgeJ1mbo)

# homebridge-APC-Back-UPS-HS500

A HomeBridge interface for APC's 'Structured Cabling' Back-UPS HS500 UPS device, capable of monitoring, controlling and configuring the UPS.

Key features:

- Monitoring the UPS mains inlet via Contact Sensor accessory, which can generate alerts. If the UPS switches to battery power, the Contact Sensor is reported as 'open' and you will receive a notification in iOS on a phone or watch.
- Monitoring of the battery charge level and charging status through Battery Service. Access this through the Contact Sensor settings within the Home app.
- Reporting of UPS load and run-time, which is done through Light Sensor accessories as there isn't really any plain 'gauge' we can add in HomeKit, and custom properties are not shown in the Home app.
- Control of the three UPS outputs - these are exposed as Outlet accessories (see screenshot below). This control can also be disabled individually, to prevent accidental power-off for example or a core switch or similar.
- V3 adds support to power-cycle outlets (i.e., forcibly reboot attached equipment).
- Automatic configuration of UPS management interface for static IP or DHCP. If you've never managed to configure the HS500 network interface because APCs lousy tool only ran on Windows 98/2000, this plugin can do it for you as it initialises.
- Auto-discovery of configured UPS network address (provided only one HS500 is on the network).

<img src="https://user-images.githubusercontent.com/784541/83963920-34f69e00-a8a1-11ea-9d1b-cb2c673c23e2.png" width="250"/>
<img src="https://user-images.githubusercontent.com/784541/84699513-8a8a1500-af49-11ea-8dc1-f507a912a9e1.png" width="250"/>
<img src="https://user-images.githubusercontent.com/784541/84699747-fb313180-af49-11ea-85a5-b80c2a01cb1e.png" width="250"/>

Credits: This plugin was developed using cr3ative/homebridge-apcaccess as a starting point.

Subscribable events:

- Contact State
- Low Battery

# Plugin Configuration

Installed through HomeBridge plugins UI, the settings are fully configurable in the UI (see screenshot above).

# Issues and Contact

Please raise an issue should you come across one via Github.



CHANGELOG.md
# [3.0.2]

*** HomeBridge plugin for APC's HS500 'Structured Cabling' UPS ***

Thank you for trying this plugin! It enables monitoring and control of APC's HS500 UPS from HomeKit, as well as automatic management interface configuration.

Bugfixes:
3.0.2 - Internal control variable bugfix
3.0.1 - Fixed reboot request always being applied to Outlet 3
2.0.3 - Added option to make a donation through Homebridge UI
2.0.2 - README update to reflect verified status
2.0.1 - fixes for spurious validation messages on startup.

Version 3:
- Added support for UPS 'reboot' function - which power-cycles an outlet. This makes it possible to safely power-cycle an outlet even when HomeBridge is itself powered from that outlet (in which case, Homebridge will also be power-cycled).
- Added support to display embedded accessory names correctly on iOS 16/17

Version 2:
Adds support for:
- automatic configuration of UPS management interface for static IP or DHCP
- auto-discovery of UPS network address
- control of the three UPS outputs (by exposing them as outlet devices)
- allows the blocking of UPS outlet control to prevent accidental power-off
- reporting of UPS load and run-time (through lightsensor devices)

Version 1:
Provides:
- monitoring of the UPS mains inlet via Contact Sensor
- monitoring of the battery charge level and charging status through Battery Service.

If you have any feedback or wish to contribute or extend, please log an issue on the GitHub project page.

Depends on curl and, for auto discovery and configuration, net-tools, apring, xxd and socat. Install these packages in your OS (debian, ubuntu, raspbian) using:

sudo apt install curl, net-tools, arping, xxd, socat


apc500.sh
#!/bin/bash
# UPS Management script for APC 500 HS
# Depends on (sudo apt install) net-tools, apring, xxd, socat
# Will work only when precised one APC 500 HS is connected to the LAN.
#
# -f - to Discover an HS500 device on the LAN and show its configuration
# -s=[0.0.0.0] - to Set the HS500 management IP address (enter 0.0.0.0 for DHCP) and show its resulting configuration
# -n=[name] - to set the HS500 Name and show its resulting configuration
# -i=[eth0] - to specify the interface to use (defaults to eth0)
#

# DEFAULTS
INTERFACE="eth0"
SETNAMERESULT="Did not run"
SETIPRESULT="Did not run"

# Command line parameters - what are we doing?
for i in "$@"
do
  case $i in
      -f*|--find*)
        FUNCTION="FIND"
        ;;
      -s=*|--setip=*)
        FUNCTION="SETIP"
        IPADDRESS="${i#*=}"
        shift # past argument=value
        ;;
      -n=*|--setname=*)
        FUNCTION="SETNAME"
        NAME="${i#*=}"
        shift # past argument=value
        ;;
      -i=*|--interface=*)
        INTERFACE="${i#*=}"
        shift # past argument=value
        ;;
      *)
        # unknown option
        ;;
  esac
done


if [ "$FUNCTION" = "SETIP" ]; then
  # Work out IP in Hex
  IPDEC="$(echo "$IPADDRESS" | sed 's/\./ /g')"
  TMP="$(printf '%02x ' $IPDEC ; echo)"
  IPHEX=${TMP::-1}
  RESULT="$(echo '12 50 00 a0 98 05 45 43 f5 f4 34 f6 '"$IPHEX" | xxd -r -p | socat - UDP4-DATAGRAM:255.255.255.255:9950,so-broadcast | tr -dc '[:print:]\n')"
  case "$RESULT" in
    "") SETIPRESULT="UPS did not respond.";;
    *) SETIPRESULT="UPS acknowledged command."
  esac
  # sleep for 1 second to allow the management interface to process the changes and acquire a DHCP address is so configured
  # sleep 1
fi;


if [ "$FUNCTION" = "SETNAME" ]; then
  TMP="$(echo "$NAME" | xxd -p )"
  NAMEHEX=${TMP::-2}
  RESULT="$(echo '12 50 00 a0 10 08 45 43 f5 '"$NAMEHEX"' 00' | xxd -r -p | socat - UDP4-DATAGRAM:255.255.255.255:9950,so-broadcast | tr -dc '[:print:]\n')"
  case "$RESULT" in
    "") SETNAMERESULT="UPS did not respond.";;
    *) SETNAMERESULT="UPS acknowledged command."
  esac
  # sleep for 1 second to allow the management interface to process the change
  # sleep 1
fi;


# we always run this part, to display the active configuration
#if [ "$FUNCTION" = "FIND" ]; then
  # Find UPS via broadcast
  DATA="$(echo '11 50 00 A0 10 50 43 43' | xxd -r -p | socat - UDP4-DATAGRAM:255.255.255.255:9950,so-broadcast,sourceport=9951 | tr -dc '[:print:]\n')"
  MODEL=${DATA:9:15}
  SERIAL=${DATA:24:12}
  MACADD=${DATA:37:12}
  TMP=${MACADD,,}
  MACADDR=${TMP:0:2}:${TMP:2:2}:${TMP:4:2}:${TMP:6:2}:${TMP:8:2}:${TMP:10:2}
  TMP=${DATA:52}
  NAME=${TMP::-1}
  # And lookup the MAC from ARP cache
  # IPADD="$(arp -an | grep "$MACADDR" | egrep -o '[0-9]*\.[0-9]*\.[0-9]*\.[0-9]*')"
  # Likely, there was nothing there. Check, and use arping (needs root) if we need to
  # case "$IPADD" in
  #  "")
      IPADD="$(arping "$MACADDR" -c 2 -i "$INTERFACE" | egrep -o -m 1 '[0-9]*\.[0-9]*\.[0-9]*\.[0-9]*' | head -1)"
  #     ;;
  # esac

  # Display the collected device configuration in json format
  echo '{'
  echo '  "Model":"'$MODEL'",'
  echo '  "SerialNumber":"'$SERIAL'",'
  echo '  "Name":"'$NAME'",'
  echo '  "MACAddress":"'$MACADDR'",'
  echo '  "SetIpResult":"'$SETIPRESULT'",'
  echo '  "SetNameResult":"'$SETNAMERESULT'",'
  # Now check again to see if we have an IP address. If we do, we can get run-time information, eg battery level etc
  case "$IPADD" in
    "")
      echo '  "IPAddress":"0.0.0.0"'
      ;;
    *)
      echo '  "IPAddress":"'$IPADD'"'
    esac
  echo '}'
#fi;


# End of script.



check-power-xml.sh
#!/bin/bash
# This script uses apc.sh to check the status of BackUPS HS-500
# And outputs the status as an XML record

STATUS="/tmp/apc-500-status.tmp"				# temp file for operational status - battery level etc
OUTPUTFILE="/tmp/apc500status.xml"
UPS=$1

# get status values from the primitive web UI
# logger "Getting UPS Status from $UPS"
curl -sl "http://$UPS/status.cgi" | tr -dc '[:print:]\n' > $STATUS

# Extract the unit operating status fields - battery level etc
LOAD="$(cat $STATUS | grep -o '[0-9]*&nbsp;Watts' | grep -o '[0-9]*')"
BATTERYLEVEL="$(cat $STATUS | grep -o '[0-9]*&nbsp;%' | grep -o '[0-9]*')"
RUNTIME="$(cat $STATUS | grep -o '[0-9]*&nbsp;minutes' | grep -o '[0-9]*')"
BATTERYSTATUS="$(cat $STATUS | egrep -o 'Charged|Charging|Discharged|Discharging')"
UPSSTATUS="$(cat $STATUS | egrep -o 'On&nbsp;Line|On&nbsp;Battery' | sed 's/&nbsp;/ /g')"
LASTTEST="$(cat $STATUS | egrep -o 'Result of last self-test is:.*(Passed|Failed)</font>' | egrep -o '(Passed|Failed)')"
LASTTRANSFER="$(cat $STATUS | egrep -o 'No&nbsp;Transfer|Blackout' | sed 's/&nbsp;/ /g')"

# log process
# logger "UPS Status $UPSSTATUS, $RUNTIME minutes remaining (load: $LOAD Watts)"

# Generate XML
echo '<?xml version="1.0" encoding="UTF-8" ?>'
echo '<root>'
echo '  <0>'
echo '    <UPSSTATUS>'$UPSSTATUS'</UPSSTATUS>'
echo '    <LOAD>'$LOAD'</LOAD>'
echo '    <BATTERYLEVEL>'$BATTERYLEVEL'</BATTERYLEVEL>'
echo '    <BATTERYSTATUS>'$BATTERYSTATUS'</BATTERYSTATUS>'
echo '    <RUNTIME>'$RUNTIME'</RUNTIME>'
echo '    <LASTTEST>'$LASTTEST'</LASTTEST>'
echo '    <LASTTRANSFER>'$LASTTRANSFER'</LASTTRANSFER>'
echo '  </0>'
echo '</root>'

# garbage collector
rm -f $STATUS


check-power-json.sh
#!/bin/bash
# This script uses apc.sh to check the status of BackUPS HS-500
# And outputs the status as an json record

STATUS="/tmp/apc-500-status.tmp"
UPS=$1

# get status values from the primitive web UI
# logger "Getting UPS Status from $UPS"
curl -sl "http://$UPS/status.cgi" | tr -dc '[:print:]\n' > $STATUS

# Extract the unit operating status fields - battery level etc
LOAD="$(cat $STATUS | grep -o '[0-9]*&nbsp;Watts' | grep -o '[0-9]*')"
BATTERYLEVEL="$(cat $STATUS | grep -o '[0-9]*&nbsp;%' | grep -o '[0-9]*')"
RUNTIME="$(cat $STATUS | grep -o '[0-9]*&nbsp;minutes' | grep -o '[0-9]*')"
BATTERYSTATUS="$(cat $STATUS | egrep -o 'Charged|Charging|Discharged|Discharging')"
UPSSTATUS="$(cat $STATUS | egrep -o 'On&nbsp;Line|On&nbsp;Battery' | sed 's/&nbsp;/ /g')"
LASTTEST="$(cat $STATUS | egrep -o 'Result of last self-test is:.*(Passed|Failed)</font>' | egrep -o '(Passed|Failed)')"
LASTTRANSFER="$(cat $STATUS | egrep -o 'No&nbsp;Transfer|Blackout' | sed 's/&nbsp;/ /g')"

# log process
# logger "UPS Status $UPSSTATUS, $RUNTIME minutes remaining (load: $LOAD Watts)"

# Generate json
echo '{'
echo '    "upsstatus":"'$UPSSTATUS'",'
echo '    "load":'$LOAD','
echo '    "batterylevel":'$BATTERYLEVEL','
echo '    "batterystatus":"'$BATTERYSTATUS'",'
echo '    "runtime":'$RUNTIME','
echo '    "lasttest":"'$LASTTEST'",'
echo '    "lasttransfer":"'$LASTTRANSFER'"'
echo '}'

# garbage collector
rm -f $STATUS


apc.sh
#!/bin/bash
# This script allows you to control APC BACK-UPS HS 500 from command line;
# Please install cURL before using; script works with web-interface of UPS;
# Copyright (c) 2011-2015 Anton Bagayev, abagayev@gmail.com, http://dontgiveafish.com
#
# Updated by James Pearce, June 2020, to report status as JSON block, to enable it's use in 
# HomeBridge integration plugin: https://github.com/J1mbo/homebridge-APC-Back-UPS-HS500
#

COOKIES=$(mktemp)					# cookie temp file
CFG=$(mktemp)						# temp file to get current configuration
STATUS=$(mktemp)					# temp file for operational status - battery level etc
# automatically delete the tmp files on exit...
trap "{ rm -f $COOKIES; rm -f $CFG; rm -f $STATUS; }" EXIT

function onoffreboot {
# this function will help us to humanize statuses and options
# we use it this way: onoffreboot $1
# where $1 is number or humanized word  (for ex, 2=reboot)

        case "$1" in
	on)	echo 0;;
	off)	echo 1;;
	reboot)	echo 2;;
	0)	echo on;;
	1)	echo off;;
	2)	echo reboot;;
	toggle)	echo toggle;;
	esac

}

function outputvalue {
# this function will help us get to know current configuration
# use it this way: outputvalue $1 $2 
# where  $1 is filename, $2 is output var (for ex, q is var for output3)

	if [ $(grep -ca "Checked..name=$2 value=0" $1) = 1 ]; then echo 0; fi
	if [ $(grep -ca "Checked..name=$2 value=1" $1) = 1 ]; then echo 1; fi
	if [ $(grep -ca "Checked..name=$2 value=2" $1) = 1 ]; then echo 2; fi
}

function toggle {
# this function will help us to toggle configuration
# use it this way: toggle $1
# where $1 is configuration(0 or 1)

        case "$1" in
	0)	echo 1;;
	1)	echo 0;;
	*)	echo $1;;
	esac
}

# parse options
# show help and exit if nothing
if [ -e $2 ]; then {
	echo "Usage: "
	echo "  apc.sh ip=[ip-addres] status";
	echo "or"
	echo "  apc.sh [output1=on|off|toggle|reboot] [output2=on|off|toggle|reboot] [output3=on|off|toggle|reboot] user=[username] pass=[passcode]";
	echo ""
	echo "Note: [passcode] is the apc encoded password and must not include any digits (characters only)"
	exit;
}
fi

# parse query
for arg in "$@"; do
  case "$arg" in
	ip=*)			UPS="$(echo $arg | cut -d'=' -f2)";;
	output1=*)		o1=$(onoffreboot `echo $arg | cut -d'=' -f2`);;
	output2=*)		o2=$(onoffreboot `echo $arg | cut -d'=' -f2`);;
	output3=*)		o3=$(onoffreboot `echo $arg | cut -d'=' -f2`);;
	status)			status=TRUE;;
	user=*)			USERNAME="$(echo $arg | cut -d'=' -f2)";;
	pass=*)			PASSWORD="$(echo $arg | cut -d'=' -f2)";;
  esac
done

## toggle values if toggle is in query
#if [ "$o1" = "toggle" ]; then o1=$(toggle $(outputvalue $CFG o)); fi
#if [ "$o2" = "toggle" ]; then o2=$(toggle $(outputvalue $CFG p)); fi
#if [ "$o3" = "toggle" ]; then o3=$(toggle $(outputvalue $CFG q)); fi

# get current oulet states
curl -sl "http://$UPS/CFG1.CGI" > $CFG
if [ -e $o1 ]; then o1=$(outputvalue $CFG o); fi
if [ -e $o2 ]; then o2=$(outputvalue $CFG p); fi
if [ -e $o3 ]; then o3=$(outputvalue $CFG q); fi

if [ -n "$status" ]; then
	# status returns the UPS status only
	# get UPS status page from web-control
	curl -sl "http://$UPS/status.cgi" | tr -dc '[:print:]\n' > $STATUS

	# Extract the unit operating status fields - battery level etc
	LOAD="$(cat $STATUS | grep -o '[0-9]*&nbsp;Watts' | grep -o '[0-9]*')"
	BATTERYLEVEL="$(cat $STATUS | grep -o '[0-9]*&nbsp;%' | grep -o '[0-9]*')"
	RUNTIME="$(cat $STATUS | grep -o '[0-9]*&nbsp;minutes' | grep -o '[0-9]*')"
	BATTERYSTATUS="$(cat $STATUS | egrep -o 'Charged|Charging|Discharged|Discharging')"
	UPSSTATUS="$(cat $STATUS | egrep -o 'On&nbsp;Line|On&nbsp;Battery' | sed 's/&nbsp;/ /g')"
	LASTTEST="$(cat $STATUS | egrep -o 'Result of last self-test is:.*(Passed|Failed)</font>' | egrep -o '(Passed|Failed)')"
	LASTTRANSFER="$(cat $STATUS | egrep -o 'No&nbsp;Transfer|Blackout' | sed 's/&nbsp;/ /g')"

        # Default values to provide valid json object if nothing is returned from UPS
	[ -z "$LOAD" ] && LOAD=0
	[ -z "$BATTERYLEVEL" ] && BATTERYLEVEL=0
	[ -z "$RUNTIME" ] && RUNTIME=0

	echo '{'
	echo '    "output1":"'$(onoffreboot $o1)'",'
        echo '    "output2":"'$(onoffreboot $o2)'",'
        echo '    "output3":"'$(onoffreboot $o3)'",'
	echo '    "upsstatus":"'$UPSSTATUS'",'
	echo '    "load":'$LOAD','
	echo '    "batterylevel":'$BATTERYLEVEL','
        echo '    "batterystatus":"'$BATTERYSTATUS'",'
	echo '    "runtime":'$RUNTIME','
	echo '    "lasttest":"'$LASTTEST'",'
	echo '    "lasttransfer":"'$LASTTRANSFER'"'
	echo '}'
else
	# power outlet control requested...
	# login first
	curl -slo /dev/null --cookie $COOKIES "http://$UPS/2?n=$USERNAME&T=$PASSWORD"
	# install configuration
	curl -slo /dev/null --cookie $COOKIES "http://$UPS/3?s=1&a=2&u=10&l=16&o=$o1&p=$o2&q=$o3&S2=Apply"
	# logout
	curl -slo /dev/null --cookie $COOKIES "http://$UPS/Logon.cgi"
fi

# garbage collector - remove temp files
#rm -f $CFG
#rm -f $COOKIES
#rm -f $STATUS
